core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -140,12 +159,18 @@ public void write(T row) throws IOException {
       PathOffset previous = insertedRowMap.put(copiedKey, pathOffset);
       if (previous != null) {
         // TODO attach the previous row if has a positional-delete row schema in appender factory.
-        posDeleteWriter.delete(previous.path, previous.rowOffset, null);
+        writePosDelete(previous);
       }
 
       dataWriter.write(row);
     }
 
+    private void writePosDelete(PathOffset pathOffset) {
+      PositionDelete<T> delete = PositionDelete.create();
```

Comment:
We usually avoid creating an extra `PositionDelete` object for every delete, if possible. Instead, we initialize one object and reuse it. Take a look at `BasePositionDeltaWriter`.

Reply:
Done